torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1164,12 +1164,60 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::alltoall_base(
 }
 #endif
 
+#ifdef ENABLE_NCCL_P2P_SUPPORT
+std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::alltoall(
+    std::vector<at::Tensor>& outputTensors,
+    std::vector<at::Tensor>& inputTensors,
+    const AllToAllOptions& /* unused */) {
+  auto device = outputTensors[0].device();
+  for (size_t r = 0; r < outputTensors.size(); r++) {
+    check_gpu_single_tensor(outputTensors[r]);
+    check_gpu_single_tensor(inputTensors[r]);
+    TORCH_CHECK(device == outputTensors[r].device() && device == inputTensors[r].device(),
+      "Tensors must be on the same device")
+  }
+  std::vector<at::Tensor> inputTensor0 = {inputTensors[0]};
+  std::vector<at::Tensor> outputTensor0 = {outputTensors[0]};
+  return collective(
+    inputTensor0,
+    outputTensor0,
+    [&](at::Tensor& /* unused */,
+        at::Tensor& /* unused */,
+        ncclComm_t comm,
+        at::cuda::CUDAStream& stream) {
+      for (size_t r = 0; r < outputTensors.size(); r++) {
```

Comment:
Why not reuse NcclAlltoallv to do the send/recv? 
https://github.com/pytorch/pytorch/blob/37093f4d99c22cb9d89d9eb1cd56b97be92938b4/torch/lib/c10d/ProcessGroupNCCL.cpp#L187

Reply:
`ncclAlltoallv` requires the arguments to be in a format of base pointers, displacements, and counts. This is good for `all_to_all_single` because we are sending/receiving slices of tensors. But for `all_to_all`, since each tensor in the list independent, converting the arguments into this format is not easier than implementing it with ncclSend/ncclRecv directly.

Reply:
@zasdfgbnm thanks for you patience and sorry for the delay. 

One minor change: put the NCCL Send/Recv with a group start/end similar to here: https://github.com/pytorch/pytorch/blob/92921c82bb7c0d4f1874ab354d2aa903b662c355/torch/lib/c10d/ProcessGroupNCCL.cpp#L183

This should not have effect on performance since these operations get encapsulated in a group start/end in the collective: 
https://github.com/pytorch/pytorch/blob/92921c82bb7c0d4f1874ab354d2aa903b662c355/torch/lib/c10d/ProcessGroupNCCL.cpp#L1065

The reason for requesting this change is to give the clear impression that all these sends/recvs are notionally part of a single call.

Reply:
@srinivas212 Done!